File: supabase/functions/_shared/scoring.parity.test.ts
// Parity test — TS side of the contract enforced by shared/parity-fixture.json.
// The same fixture is read by shared/.../ScoringParityTest.kt (Kotlin side).
// Run: deno test --allow-read supabase/functions/_shared/scoring.parity.test.ts

import { assertAlmostEquals, assertEquals } from 'https://deno.land/std@0.224.0/assert/mod.ts';
import { distanceScore, realTimeRelevance, totalScore } from './scoring.ts';
import { heuristicVersion } from '../../../shared/types.ts';

interface Fixture {
  heuristicVersion: string;
  inputs: {
    walkMinutes: number;
    breakdownBase: { consensus: number; recency: number; useCaseFit: number };
    event: {
      startISO: string;
      endISO: string;
      announcedAtISO: string;
      nowISO: string;
      isAtThisVenue: boolean;
      isPurposeAligned: boolean;
    };
  };
  expected: {
    distance: number;
    realTimeRelevance: number;
    totalScore: number;
  };
}

async function loadFixture(): Promise<Fixture> {
  // Resolve relative to this test file so it works regardless of cwd.
  const url = new URL('../../../shared/parity-fixture.json', import.meta.url);
  return JSON.parse(await Deno.readTextFile(url));
}

Deno.test('parity: heuristicVersion constant equals fixture', async () => {
  const fx = await loadFixture();
  assertEquals(heuristicVersion, fx.heuristicVersion);
});

Deno.test('parity: distanceScore matches fixture', async () => {
  const fx = await loadFixture();
  assertEquals(distanceScore(fx.inputs.walkMinutes), fx.expected.distance);
});

Deno.test('parity: realTimeRelevance matches fixture', async () => {
  const fx = await loadFixture();
  const e = fx.inputs.event;
  const rtr = realTimeRelevance(
    e.startISO,
    e.endISO,
    e.announcedAtISO,
    e.nowISO,
    e.isAtThisVenue,
    e.isPurposeAligned,
  );
  assertAlmostEquals(rtr, fx.expected.realTimeRelevance, 1e-9);
});

Deno.test('parity: totalScore (5-signal) matches fixture', async () => {
  const fx = await loadFixture();
  const e = fx.inputs.event;
  const distance = distanceScore(fx.inputs.walkMinutes);
  const rtr = realTimeRelevance(
    e.startISO,
    e.endISO,
    e.announcedAtISO,
    e.nowISO,
    e.isAtThisVenue,
    e.isPurposeAligned,
  );
  const total = totalScore({
    ...fx.inputs.breakdownBase,
    distance,
    realTimeRelevance: rtr,
  });
  assertAlmostEquals(total, fx.expected.totalScore, 1e-9);
});

File: shared/src/commonTest/kotlin/com/austinconnor/pulse/shared/ScoringParityTest.kt
package com.austinconnor.pulse.shared

import kotlin.test.Test
import kotlin.test.assertEquals

// Kotlin side of the TS<->Kotlin scoring-parity contract.
// The TS side reads shared/parity-fixture.json directly via Deno.readTextFile in
// scoring.parity.test.ts. commonTest resource-loading is target-specific in KMP, so
// here we inline the same fixture values as constants. ANY change here must mirror
// shared/parity-fixture.json byte-for-byte (verified by ParityFixtureSnapshotTest below).

private object Fixture {
    const val HEURISTIC_VERSION = "5sig-2026.06"

    // inputs
    const val WALK_MINUTES = 12
    const val CONSENSUS = 1.5
    const val RECENCY = 1.2
    const val USE_CASE_FIT = 1.0
    const val EVENT_START_ISO = "2026-06-12T20:00:00Z"
    const val EVENT_END_ISO = "2026-06-12T23:00:00Z"
    const val ANNOUNCED_AT_ISO = "2026-06-12T08:00:00Z"
    const val NOW_ISO = "2026-06-12T19:30:00Z"
    const val IS_AT_THIS_VENUE = true
    const val IS_PURPOSE_ALIGNED = true

    // expected
    const val EXPECTED_DISTANCE = 1
    const val EXPECTED_REAL_TIME_RELEVANCE = 3.0
    const val EXPECTED_TOTAL_SCORE = 7.7
}

private const val EPSILON = 1e-9

class ScoringParityTest {

    @Test
    fun heuristicVersion_matches_fixture() {
        assertEquals(Fixture.HEURISTIC_VERSION, heuristicVersion)
    }

    @Test
    fun distanceScore_matches_fixture() {
        assertEquals(Fixture.EXPECTED_DISTANCE, Scoring.distanceScore(Fixture.WALK_MINUTES))
    }

    @Test
    fun realTimeRelevance_matches_fixture() {
        val rtr = Scoring.realTimeRelevance(
            eventStartISO = Fixture.EVENT_START_ISO,
            eventEndISO = Fixture.EVENT_END_ISO,
            announcedAtISO = Fixture.ANNOUNCED_AT_ISO,
            nowISO = Fixture.NOW_ISO,
            isAtThisVenue = Fixture.IS_AT_THIS_VENUE,
            isPurposeAligned = Fixture.IS_PURPOSE_ALIGNED,
        )
        assertEquals(Fixture.EXPECTED_REAL_TIME_RELEVANCE, rtr, EPSILON)
    }

    @Test
    fun totalScore_5signal_matches_fixture() {
        val rtr = Scoring.realTimeRelevance(
            eventStartISO = Fixture.EVENT_START_ISO,
            eventEndISO = Fixture.EVENT_END_ISO,
            announcedAtISO = Fixture.ANNOUNCED_AT_ISO,
            nowISO = Fixture.NOW_ISO,
            isAtThisVenue = Fixture.IS_AT_THIS_VENUE,
            isPurposeAligned = Fixture.IS_PURPOSE_ALIGNED,
        )
        val breakdown = ScoreBreakdown(
            consensus = Fixture.CONSENSUS,
            recency = Fixture.RECENCY,
            useCaseFit = Fixture.USE_CASE_FIT,
            distance = Scoring.distanceScore(Fixture.WALK_MINUTES).toDouble(),
            realTimeRelevance = rtr,
        )
        assertEquals(Fixture.EXPECTED_TOTAL_SCORE, Scoring.totalScore(breakdown), EPSILON)
    }
}
